main.py

```diff
@@ -0,0 +1,449 @@
+                    local_region = valid_depth_heightmap[max(best_pix_y - safe_kernel_width, 0):min(best_pix_y + safe_kernel_width + 1, valid_depth_heightmap.shape[0]), max(best_pix_x - safe_kernel_width, 0):min(best_pix_x + safe_kernel_width + 1, valid_depth_heightmap.shape[1])]
+                    if local_region.size == 0:
+                        safe_z_position = workspace_limits[2][0]
+                    else:
+                        safe_z_position = np.max(local_region) + workspace_limits[2][0]
+                    primitive_position[2] = safe_z_position
+
+                # Save executed primitive
+                if nonlocal_variables['primitive_action'] == 'push':
+                    trainer.executed_action_log.append([0, nonlocal_variables['best_pix_ind'][0], nonlocal_variables['best_pix_ind'][1], nonlocal_variables['best_pix_ind'][2]]) # 0 - push
+                elif nonlocal_variables['primitive_action'] == 'grasp':
+                    trainer.executed_action_log.append([1, nonlocal_variables['best_pix_ind'][0], nonlocal_variables['best_pix_ind'][1], nonlocal_variables['best_pix_ind'][2]]) # 1 - grasp
+                logger.write_to_log('executed-action', trainer.executed_action_log)
+
+                # Visualize executed primitive, and affordances
+                if save_visualizations:
+                    push_pred_vis = trainer.get_prediction_vis(push_predictions, color_heightmap, nonlocal_variables['best_pix_ind'])
+                    logger.save_visualizations(trainer.iteration, push_pred_vis, 'push')
+                    cv2.imwrite('visualization.push.png', push_pred_vis)
+                    grasp_pred_vis = trainer.get_prediction_vis(grasp_predictions, color_heightmap, nonlocal_variables['best_pix_ind'])
+                    logger.save_visualizations(trainer.iteration, grasp_pred_vis, 'grasp')
+                    cv2.imwrite('visualization.grasp.png', grasp_pred_vis)
+
+                # Initialize variables that influence reward
+                nonlocal_variables['push_success'] = False
+                nonlocal_variables['grasp_success'] = False
+                change_detected = False
+
+                # Execute primitive
+                if nonlocal_variables['primitive_action'] == 'push':
+                    nonlocal_variables['push_success'] = robot.push(primitive_position, best_rotation_angle, workspace_limits)
+                    print('Push successful: %r' % (nonlocal_variables['push_success']))
+                elif nonlocal_variables['primitive_action'] == 'grasp':
+                    nonlocal_variables['grasp_success'] = robot.grasp(primitive_position, best_rotation_angle, workspace_limits)
+                    print('Grasp successful: %r' % (nonlocal_variables['grasp_success']))
+
+                nonlocal_variables['executing_action'] = False
+
+            time.sleep(0.01)
+    action_thread = threading.Thread(target=process_actions)
+    action_thread.daemon = True
```

Comment:
A daemon thread does not block the main thread from exiting and continues to run in the background.